corehq/apps/reports/util.py

```diff
@@ -641,6 +642,17 @@ def _get_hq_group_id(session):
     return session.get_group(HQ_TABLEAU_GROUP_NAME).get('id')
 
 
+def clean_tableau_role(role_name):
```

Comment:
Doesn't setting `choices` to `TABLEAU_ROLES` already do this validation if `obj.clean()` is called?
```python
role = models.CharField(max_length=32, choices=TABLEAU_ROLES)
```

You can also pass `validators` to the field as described here:
https://docs.djangoproject.com/en/5.0/ref/validators/

Reply:
>Doesn't setting choices to TABLEAU_ROLES already do this validation if obj.clean() is called?

My understanding was that `clean` needed to be implemented and otherwise didn't do any validation
. "This method should be used to provide custom model validation, and to modify attributes on your model if desired."

I like the idea of passing the validator to the field.

Reply:
Ah, looks like it's `full_clean()`:
https://docs.djangoproject.com/en/4.2/ref/models/instances/#django.db.models.Model.full_clean
> This method calls Model.clean_fields(), Model.clean(), Model.validate_unique(), and Model.validate_constraints() 

> Note that full_clean() will not be called automatically when you call your model’s save() method. You’ll need to call it manually when you want to run one-step model validation for your own manually created models. For example:

I think that should apply the choices validation even without anything custom being added.

Reply:
Just saw your other message about problems with `full_clean` - that's frustrating.  The problem there is that we've defined the model incorrectly, unfortunately.  Might be worth adding `blank=True` to all the nullable fields - I don't think that actually runs a real migration in the db (though I think it will make a new migration file - not certain why that is).

If not, I think you can also do a clumsy targeted validation:
```python
if self.tableau_role:
    field = self._meta.get_field('tableau_role')
    field.validate(self.tableau_role, self)
```

Reply:
Thank you for that, that's great news.  I think this brings more order to things rather than adding complexity, which is always satisfying when possible.